internal/configs/virtualserver.go

```diff
@@ -841,6 +841,27 @@ func generateProxyPassProtocol(enableTLS bool) string {
 	return "http"
 }
 
+func generateGrpcPass(grpc bool, tlsEnabled bool, upstreamName string, internal bool) string {
+	proxyPass := fmt.Sprintf("%v://%v", generateGrpcPassProtocol(tlsEnabled), upstreamName)
+
+	if !grpc {
+		return ""
+	}
+
+	if internal {
+		return fmt.Sprintf("%v$request_uri", proxyPass)
```

Comment:
unfortunately, this will not work for gRPC. You can check it if you deploy a VS like this:
```yaml
apiVersion: k8s.nginx.org/v1
kind: VirtualServer
metadata:
  name: grpc-example
spec:
  host: grpc.example.com
  tls:
    secret: grpc-secret
    redirect:
      enable: true
  upstreams:
  - name: grpc-app 
    service: grpc-svc
    port: 50051 
    # grpc: true
  - name: grpc-app-2 
    service: grpc-svc
    port: 50051 
    # grpc: true
  routes:
  - path: /
    splits:
    - weight: 50
      action:
        pass: grpc-app
        grpc: true
    - weight: 50
      action:
        pass: grpc-app-2
        grpc: true
```

In the NGINX logs you will see something simliar, depending on your app:
```
2020/09/11 02:56:07 [error] 71#71: *62 invalid host in upstream "vs_default_grpc-example_grpc-app-2/helloworld.Greeter/SayHello", client: 127.0.0.1, server: grpc.example.com, request: "POST /helloworld.
Greeter/SayHello HTTP/2.0", host: "grpc.example.com:8443"
127.0.0.1 - - [11/Sep/2020:02:56:07 +0000] "POST /helloworld.Greeter/SayHello HTTP/2.0" 500 1 "-" "grpc-go/1.29.0-dev" "-"
```

There is a solution to fix that: instead of doing `grpc_pass grpc://vs_default_grpc-example_grpc-app-2$request_uri;` in the Config, do `rewrite ^ $request_uri break;` . That will work :) 

To do that,  please update `generateRewrites` function, so that it inserts `rewrite ^ $request_uri break;` if the location is internal and gPRC is enabled for the upstream. Note that the function might add additional configured rewrites, so  in that case `rewrite ^ $request_uri break;` will not have `break` in the end.

Reply:
Thanks for the detailed explanation, it seems we already have rewrite rule for internal, correct me if I'm wrong, is it means we don't need additional rewrite rule?

https://github.com/nginxinc/kubernetes-ingress/blob/1ee2bcdfa033278a9ad2ba534072cd167b8b8a1f/internal/configs/virtualserver.go#L1186